### PR TITLE
Update url.move

### PR DESCRIPTION
**Pull request by benr-ml**

### CODE REVIEW

**Comment by benr-ml** [comment]
@kchalkias, I took a first stab, please let me know if that's the direction you had in mind

File: crates/sui-framework/sources/external_resource.move [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    const HASH_VECTOR_LENGTH: u64 = 32;
+
+    /// Error code when the length of the digest vector is not HASH_VECTOR_LENGTH.
+    const EHashLengthMismatch: u64 = 0;
+
+    /// URL: standard Uniform Resource Locator string.
+    ///
+    /// MUST follow RFC-3986.
+    /// Clients MUST support (at least) the following schemes: ipfs, https.
+    struct Url has store, copy, drop {
```
Comment (sblackshear) [inline crates/sui-framework/sources/external_resource.move:25]: I think this may want to continue living in a separate `url.move` file, both to avoid breaking changes for clients and for more idiomatic Move style (typically, each module should have at most one "main" type).

**Comment by sblackshear** [comment]
This looks great to me.

File: crates/sui-framework/sources/digest.move [hunk 2]
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Sui types for message digests.
+module sui::digest {
```
Comment (kchalkias) [inline crates/sui-framework/sources/digest.move:5]: I thought that Move supports sha256 and sha3_256 in std can we crosscheck? @tnowacki 
Also can you sync up with @joyqvq who is working on a similar hash.move module?
See: https://github.com/MystenLabs/sui/pull/5149

Comment (benr-ml) [reply crates/sui-framework/sources/digest.move:5]: Right - https://github.com/MystenLabs/sui/blob/7424aba44aece90f2969171629773bdc6c79ed7e/crates/sui-framework/deps/move-stdlib/sources/hash.move#L9-L10 - I believe that we don't want to change existing interfaces, especially of Move stdlib. New APIs should probably use a similar (new) type.

Comment (joyqvq) [reply crates/sui-framework/sources/digest.move:5]: I can wait till this lands and add keccak to digest.move?

File: crates/sui-framework/sources/digest.move [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    /// Create a `Sha256Digest` from bytes. Aborts if `bytes` is not of length 32.
+    public fun new_sha256_digest(digest: vector<u8>): Sha256Digest {
+        assert!(vector::length(&digest) == SHA256_DIGEST_VECTOR_LENGTH, EHashLengthMismatch);
+        Sha256Digest { digest }
+    }
+
+    /// Get the digest.
+    public fun sha256_digest(self: &Sha256Digest): vector<u8> {
+        self.digest
+    }
```
Comment (joyqvq) [inline crates/sui-framework/sources/digest.move:28]: perhaps calling these two methods `from_bytes` and `to_bytes` are clearer? (was thinking following the same pattern for objects in ecdsa.move to adopt the same pattern)

Comment (benr-ml) [reply crates/sui-framework/sources/digest.move:28]: done

**Review by kchalkias — COMMENTED** [review]
Note that we generally prefer rebasing vs merging to main. We will merge anyway after the PR is approved.

**Review by joyqvq — APPROVED** [review]
lgtm!

File: crates/sui-framework/sources/immutable_external_resource.move [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    /// Represents a resource that can move but must never change. Example use cases:
+    /// - NFT images.
+    /// - NFT metadata.
+    ///
+    /// `url` MUST follow RFC-3986. Clients MUST support (at least) the following schemes: ipfs, https.
+    /// `digest` MUST be set to SHA256(content of resource at `url`).
+    ///
+    /// Clients of this type MUST fetch the resource at `url`, compute its digest and compare it against `digest`. If
+    /// the result is false, clients SHOULD indicate that to users or ignore the resource.
+    struct ImmutableExternalResource has store, copy, drop {
```
Comment (kchalkias) [inline crates/sui-framework/sources/immutable_external_resource.move:24]: Before we merge we need to agree on the hash function supported. 
There are at least two reasons for this:
- (agility) i.e. IPFS links include the hash algorithm via the multi-codec format, see https://proto.school/anatomy-of-a-cid/02 and https://docs.filebase.com/ipfs/ipfs-cids (note the two CID Formats. There are currently two different versions of IPFS CIDs - version 0 (0v) and version 1 (v1).
- Sui's default hash is sha3, not sha2

TL;DR I'd prefer we take a decision on these matters in the next crypto meeting and then explain in the docs or code comments why this X or Y re supported functions.

**Review by kchalkias — CHANGES_REQUESTED** [review]

**Review by kchalkias — COMMENTED** [review]
Let's do sha3-256 as this is the current default hash function in Sui. We can iterate that later on.

**Review by kchalkias — APPROVED** [review]
Amazing job! Thanks @benr-ml
